### PR TITLE
C#: Dotnet test tracer improvements.

### PR DESCRIPTION
**Pull request by michaelnebel**
It turns out that the the tracer (unconditionally) injects the property `-p:UseSharedCompilation=false` when `dotnet test` is executed.
However, if the `dotnet test` is applied to a library or executable it results in a `dotnet` error, when the property is set is added as it is not valid in this context.

In this PR we update the LUA tracer code, such that `-p:UseSharedCompilation=false` is not injected when `dotnet test` is applied to a library or executable.

### CODE REVIEW

File: csharp/tools/tracing-config.lua [hunk 1]
```diff
@@ -48,6 +49,14 @@ function RegisterExtractorPack(id)
...
                     dotnetRunInjectionIndex = i + 1
                 end
+                if arg == 'test' then
+                    match = true
+                end
+                -- for `dotnet test`, we should not append `-p:UseSharedCompilation=false` to the command line
+                -- if a library or executable is being provided as an argument.
+                if arg:match('%.exe$') or arg:match('%.dll')  then
+                    libOrExe = true
+                end
```
Comment (hvitved) [inline csharp/tools/tracing-config.lua:59]: I guess this means that it will apply to `dotnet run` as well, which I guess is good?

Comment (michaelnebel) [reply csharp/tools/tracing-config.lua]: It would work, but it is possible to also provide the `-p:UseSharedCompilation=false` flag to `dotnet run` when it targets a `dll` or `exe` file. I will change the code such that we only skip the injection of the flag for `dotnet test`.

File: csharp/ql/lib/change-notes/2023-06-06-dotnettest.md [hunk 2]
```diff
@@ -0,0 +1,4 @@
+---
+category: minorAnalysis
+---
+* C#: Update the LUA tracer such that `-p:SharedCompilation=false` is not injected when `dotnet test` is applied to a `dll` or `exe` file.
```
Comment (mbg) [inline csharp/ql/lib/change-notes/2023-06-06-dotnettest.md:4]: I tried to reword this to focus more on the effect of the change than the implementation. How does this sound?
```suggestion
* C#: Analysis of the `dotnet test` command supplied with a `dll` or `exe` file as argument no longer fails due to the addition of an erroneous `-p:SharedCompilation=false` argument.
```

File: csharp/ql/integration-tests/posix-only/dotnet_test/test.py [hunk 3]
```diff
@@ -1,5 +1,10 @@
...
 from diagnostics_test_utils import *
 
-run_codeql_database_create(['dotnet test'], db=None, lang="csharp")
+# Implicitly build and then run tests.
+run_codeql_database_create(['dotnet test'], test_db="test-db", lang="csharp")
 check_diagnostics()
+
+# Explicitly build and then run tests.
+run_codeql_database_create(['dotnet clean', 'rm -rf test-db', 'dotnet build -o myout', 'dotnet test myout/dotnet_test.dll'], test_db="test2-db", lang="csharp")
+check_diagnostics(test_db="test2-db")
```
Comment (mbg) [inline csharp/ql/integration-tests/posix-only/dotnet_test/test.py:10]: Minor: missing new line at the end of the file?

File: csharp/tools/tracing-config.lua [hunk 4]
```diff
@@ -21,6 +21,8 @@ function RegisterExtractorPack(id)
         -- if that's `build`, we append `-p:UseSharedCompilation=false` to the command line,
         -- otherwise we do nothing.
         local match = false
+        local matchTest = false;
```
Comment (mbg) [inline csharp/tools/tracing-config.lua:24]: The variable here is named `matchTest`, but subsequently `testMatch` is used. Is that intentional or a typo?

Comment (michaelnebel) [reply csharp/tools/tracing-config.lua:24]: AUCH! I will fix this!

File: csharp/tools/tracing-config.lua [hunk 5]
```diff
@@ -48,6 +50,14 @@ function RegisterExtractorPack(id)
...
                     dotnetRunInjectionIndex = i + 1
                 end
+                if arg == 'test' then
+                    testMatch = true
+                end
+                -- for `dotnet test`, we should not append `-p:UseSharedCompilation=false` to the command line
+                -- if a library or executable is being provided as an argument.
+                if arg:match('%.exe$') or arg:match('%.dll')  then
+                    libOrExe = true
+                end
```
Comment (hvitved) [inline csharp/tools/tracing-config.lua:60]: Alternatively
```suggestion
                if testMatch and (arg:match('%.exe$') or arg:match('%.dll'))  then
                    testMatch = false
                end
```
and then you can get rid of the `libOrExe` variable.

Comment (michaelnebel) [reply csharp/tools/tracing-config.lua:60]: I also considered that (and it can be simplified even further - we don't need to check that testMatch is true in the condition. However, I was a bit unsure on the effect on the logging code on l. 40, but I will change it based on your suggestion.

File: csharp/tools/tracing-config.lua [hunk 6]
```diff
@@ -48,6 +49,14 @@ function RegisterExtractorPack(id)
                     dotnetRunNeedsSeparator = true
                     dotnetRunInjectionIndex = i + 1
                 end
+                if arg == 'test' then
+                    testMatch = true
```
[the remark this answers is not in the document]
Comment (adityasharad) [reply csharp/tools/tracing-config.lua]: So `dotnet run` when provided a dll or exe will not do any compilation?

**Review by adityasharad — COMMENTED** [review]
Some high-level suggestions.

File: csharp/tools/tracing-config.lua (hunk 6 above)
[the remark this answers is not in the document]
Comment (hvitved) [reply csharp/tools/tracing-config.lua]: Sorry, I was maybe not clear: I intended to keep the `testMatch` variable, so specifically on this line it would say
```lua
testMatch = true
match = true
```
and then keep the check `if testMatch and (arg:match('%.exe$') or arg:match('%.dll'))` instead of just `if arg:match('%.exe$') or arg:match('%.dll')`

[the remark this answers is not in the document]
Comment (michaelnebel) [reply csharp/tools/tracing-config.lua]: > So `dotnet run` when provided a dll or exe will not do any compilation?

By closer inspection, that part is actually unclear as it is not documented as a part of `dotnet run`.
We will make sure to narrow the fix such that it only applies to `dotnet test` (which was a misunderstanding between @hvitved and me)

File: csharp/tools/tracing-config.lua [hunk 7]
```diff
@@ -48,6 +49,14 @@ function RegisterExtractorPack(id)
                     dotnetRunNeedsSeparator = true
                     dotnetRunInjectionIndex = i + 1
                 end
+                if arg == 'test' then
+                    match = true
+                end
+                -- for `dotnet test`, we should not append `-p:UseSharedCompilation=false` to the command line
+                -- if a library or executable is being provided as an argument.
```
[the remark this answers is not in the document]
Comment (michaelnebel) [reply csharp/tools/tracing-config.lua:55]: Makes sense - I will extended the description and state that when an exe or dll is provided the call is forwarded to `vstest`. Thank you!

**Review by hvitved — DISMISSED** [review]
Let's do a DCA run before merging.

File: csharp/ql/integration-tests/posix-only/dotnet_test/test.py [hunk 8]
```diff
@@ -1,5 +1,10 @@
 from create_database_utils import *
 from diagnostics_test_utils import *
 
-run_codeql_database_create(['dotnet test'], db=None, lang="csharp")
+# Implicitly build and then run tests.
+run_codeql_database_create(['dotnet test'], test_db="test-db", lang="csharp")
 check_diagnostics()
+
+# Explicitly build and then run tests.
+run_codeql_database_create(['dotnet clean', 'rm -rf test-db', 'dotnet build -o myout', 'dotnet test myout/dotnet_test.dll'], test_db="test2-db", lang="csharp")
```
[the remark this answers is not in the document]
Comment (michaelnebel) [reply csharp/ql/integration-tests/posix-only/dotnet_test/test.py:9]: Yes, I will make another test.

**Review by hvitved — APPROVED** [review]

**Comment by michaelnebel** [comment]
> Let's do a DCA run before merging.

Yes, I will start it now.

File: csharp/ql/lib/change-notes/2023-06-06-dotnettest.md [hunk 9]
```diff
@@ -0,0 +1,4 @@
+---
+category: minorAnalysis
+---
+* C#: Analysis of the `dotnet test` command supplied with a `dll` or `exe` file as argument no longer fails due to the addition of an erroneous `-p:SharedCompilation=false` argument.
```
[the remark this answers is not in the document]
Comment (michaelnebel) [reply csharp/ql/lib/change-notes/2023-06-06-dotnettest.md:4]: Yes, thank you!!!
Will be fixed here: https://github.com/github/codeql/pull/13440

**Comment by smitpatel** [comment]
Which release this fix will be in? We are still seeing issue in our pipelines.

**Comment by michaelnebel** [comment]
> Which release this fix will be in? We are still seeing issue in our pipelines.

This should be in release 2.13.4. Is that the one you are using?

**Comment by smitpatel** [comment]
@michaelnebel - Thank you. I will check the version we are using and figure out upgrade path.
